lib/cdo/aws/cloudfront.rb

```diff
@@ -0,0 +1,211 @@
+    def self.create_or_update
+      cloudfront = Aws::CloudFront::Client.new
+      ids = %i(pegasus dashboard).map do |app|
+        distribution = cloudfront.list_distributions.distribution_list.items.detect do |i|
+          i.aliases.items.include?(CDO.method("#{app}_hostname").call)
+        end
+        id = nil
+        if distribution
+          id = distribution.id
+          distribution_config = cloudfront.get_distribution_config(id: id)
+          cloudfront.update_distribution(
+            id: id,
+            if_match: distribution_config.etag,
+            distribution_config: config(app, distribution_config.distribution_config.caller_reference)
+          )
+          puts "#{app} distribution updated!"
+        else
+          resp = cloudfront.create_distribution(
+            distribution_config: config(app)
+          )
+          id = resp.distribution.id
+          puts "#{app} distribution created!"
+        end
+        [app, id]
+      end
+
+      ids.map do |app, id|
+        cloudfront.wait_until(:distribution_deployed, id: id) do |waiter|
+          waiter.before_wait { |_| puts "Waiting for #{app} distribution to deploy.." }
+        end
+        puts "#{app} distribution deployed!"
+      end
+    end
+
+    # Returns a CloudFront DistributionConfig Hash compatible with the AWS SDK for Ruby v2.
+    # Syntax reference: http://docs.aws.amazon.com/sdkforruby/api/Aws/CloudFront/Types/DistributionConfig.html
+    # `app` is a symbol containing the app name (:pegasus or :dashboard)
+    def self.config(app, reference = nil)
+      config = HTTP_CACHE[app]
+      cloudfront = CONFIG[app]
+      behaviors = config[:behaviors].map do |behavior|
+        if behavior[:path].is_a? Array
+          behavior[:path].map do |path|
+            cache_behavior behavior, path
+          end
+        else
+          cache_behavior behavior, behavior[:path]
+        end
+      end.flatten
+
+      ssl_cert = cloudfront[:ssl_cert]
+      {
+        aliases: {
+          quantity: cloudfront[:aliases].length, # required
+          items: cloudfront[:aliases].empty? ? nil : cloudfront[:aliases],
+        },
+        default_root_object: '',
+        origins: {# required
+          quantity: 1, # required
+          items: [
+            {
+              id: 'cdo', # required
+              domain_name: cloudfront[:origin], # required
+              origin_path: '',
+              custom_origin_config: {
+                http_port: 80, # required
+                https_port: 443, # required
+                origin_protocol_policy: 'match-viewer', # required, accepts http-only, match-viewer
+              },
+            },
+          ],
+        },
+        default_cache_behavior: cache_behavior(config[:default]),
+        cache_behaviors: {
+          quantity: behaviors.length, # required
+          items: behaviors.empty? ? nil : behaviors,
```

Comment:
I think trial and error showed that the CloudFront API would only accept this very particular format for certain required empty lists.